src/conn_state.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "conn_state.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "context.h"
+#include "log.h"
+#include "sockaddr.h"
+#include "udp.h"
+
+static
+int
+check_preferred_route(struct enftun_conn_state* conn_state)
+{
+    struct enftun_context* ctx = (struct enftun_context*)conn_state->data;
```

Comment:
Sorry, one more change.  We shouldn't rely on the `conn_state->data` being a `struct enftun_context`.  The caller could pass anything for that data.

Since the internals of `conn_state` need access to the the `tls` socket, let's pass that in explicitly as an argument to `enftun_conn_state_start`.  And store in the `struct enftun_conn_state` as a `struct enftun_tls conn;` member (or something like that). Think of it as saying, let's start monitoring this particular TLS connection.

The other place to pass in the TLS struct would be in `enftun_conn_state_init` or `enftun_conn_state_prepare`.  But the `enftun_tls` isn't setup when we call those, so let's use `_start`.